htmd/smallmol/chemlab/builder.py

```diff
@@ -0,0 +1,406 @@
+            sm.__dict__['chiral'] = np.append(sm.chiral, a['chiral'])
+            sm.__dict__['hybridization'] = np.append(sm.hybridization, _hybridizations_IdxToType[a['hybridization']])
+            sm.__dict__['coords'] = np.concatenate((sm.coords, a['coords']), axis=0)
+            sm.__dict__['neighbors'] = np.array(sm.neighbors.tolist() + [[a['attachTo']]])
+            sm.__dict__['bondtypes'] = np.array(sm.bondtypes.tolist() + [[a['bondtype']]])
+            if isinstance(sm.neighbors[a['attachTo']], np.ndarray):
+                new_neighbors = sm.neighbors.tolist()
+                new_neighbors[a['attachTo']].append(n_atom)
+                sm.__dict__['neighbors'] = np.array(new_neighbors)
+                new_bonds = sm.bondtypes.tolist()
+                new_bonds[a['attachTo']].append(a['bondtype'])
+                sm.__dict__['bondtypes'] = np.array(new_bonds)
+                
+            else:
+                sm.neighbors[a['attachTo']].append(n_atom)
+                sm.bondtypes[a['attachTo']].append(a['bondtype'])
+
+            #TODO based on chiral
+            sm.__dict__['_chiraltags'] = np.append(sm._chiraltags, 0)
+            # Development note: removed assignement explicitHs
+
+            # coords
+            sm.__dict__['coords'][n_atom]  = self._getAtomCoords(n_atom, a['attachTo'])
+
+            #c += 1
+
+    def _getAtomCoords(self, atomIdx, attachToIdx):
+        """
+        Retruns the coordinates of the new atom based on the VSEPR theory.
+
+        Parameters
+        ----------
+        atomIdx: int
+            The index of the new atom
+        attachToIdx: int
+            The index of the atom you want to bond to
+
+        Returns
+        -------
+        newcoords: np.array
+            The np.array of the new atom coords
+        """
+        # TODO
+        # differenc hybridization from S SP SP2 SP3: Like S, P
+
+        sm =self.smallmol
+        pT = self.periodictable
+
+        atom_element = sm.element[atomIdx]
+        heavy_element = sm.element[attachToIdx]
+        heavy_coords = sm.coords[attachToIdx].reshape(3)
+        heavy_hybridization = sm.hybridization[attachToIdx]
+        heavy_neighbors = sm.neighbors[attachToIdx]
+        heavy_neighbors_num = len(heavy_neighbors)
+        bondLength = pT.getRadiusBond(heavy_element) + pT.getRadiusBond(atom_element)
+
+        dirVect = np.array([0,0,0])
+        if heavy_neighbors_num == 1:
+            dirVect[2] = 1
+
+        elif heavy_neighbors_num == 2:
+
+            nbr = self._filterNeighbors(attachToIdx, atomIdx)[0]
+            nbr_coords = sm.coords[nbr].reshape(3)
+            nbrVect = nbr_coords - heavy_coords
+
+            nbrVect = _normalizeVector(nbrVect)
+            nbrVect = -nbrVect
+
+            perpVect = _getPerpendicular(nbrVect)
+            perpVect = _normalizeVector(perpVect)
+
+            if heavy_hybridization == 4:
+                rot_matrix = _getRotationMatrix(perpVect, (180-109.471), deg=True)
```

Comment:
`180-109.471` Document such freakish numbers :P Maybe it makes sense to a chemist but I have no clue if this is a bug haha.

Reply:
This comes from the tetrahedral geometry of the VSEPR theory. SP3 has tetrahedral geometry with angles of 109.471. I have just kept as 180 - 109.471 to be clear for a chemist.